### DISCUSSION TITLE
Using Dynamic Variable/Placeholder in Rule's Filtering

### DISCUSSION THREAD

**Discussion by seinshah** [emqx/emqx#13536] [Q&A] [answered]
Hi.
Currently, I'm defining a rule to be triggered on user_connected event and eventually sending an HTTP request using some collected information from the mentioned event topic.

Rule's SQL command would look like this:
```sql
SELECT username, connected_at FROM "$events/client_connected"
```

Now, the following questions arise:

1. Is this query selects all records in the topic or just the last record?
2. What happens if another client connects between when the older client joined and rule was triggered? would the SQL result return the new client's username?
3. Is it possible to filter the SQL by a dynamic variable or placeholder such as `WHERE clientid='${clientid}'` (this format didn't work on EMQX 5.7.1)?

**Comment by zmstone** [comment, marked as answer]
Hi @seinshah 
EMQX spawns one Erlang process for each client, and you can think of rules like function calls.
This means that the rules are evaluated:
- concurrently: there is no line-up of the events pending for processing or whatsoever.
- isolated: meaning `WHERE clientid=${clientid}` would always be `true` because `clientid` is a value bind for each SQL runtime (however this would actually result in a syntax error because string interpolation or value placeholder is not supported today in rule SQL).

If you want to filter with a known client ID, you can write the client ID as a const string like this:

```
SELECT username, connected_at FROM "$events/client_connected" WHERE clientid='cid1'
```
If you have only a few such clients, you can write `and`, `or` conditions.

To write a generic rule for all clients, the client ID would have to be computed or derived from known factors or existing variable bindings.
e.g. if you want to match only the clients having a prefix, `group1-` then you can do something like:

```
SELECT username, connected_at FROM "$events/client_connected" WHERE regex_match(clientid,'^group1-.+')
```

If you have many clients to match, and the list can change at runtime?
EMQX does not natively support adding/injecting variable bindings, but it can be extended to support so.

**Reply by seinshah** [reply]
Thank you @zmstone.
I misunderstood how event topics work and now have better understanding. I though event topic is using a shared storage among all the processes which now I understand not.
